Runnable Glassfish Embedded

OndroMih opened the pull request:
Allows running GlassFish Embedded from command line: `java -jar glassfish-embedded.jar`

Supports configuration via command line arguments and/or config files in the current directory:

* set HTTP port
* deploy applications passed as arguments
* run arbitrary AsAdmin commands at boot
 
All GlassFish Embedded flavors are supported: 

* Standalone glassfish-embedded-all.jar, glassfish-embedded-web.jar, and glassfish-embedded-nucleus.jar
* glassfish-embedded-static-shell.jar in the GlassFish Server installation

Includes fixes that made GlassFish Embedded Web usable. Includes fixes to the Static shell to make it equal to standalone GlassFish Embedded JAR files.

`java -jar glassfish-embedded.jar --help` prints the following info: [glassfish.man.txt](https://github.com/user-attachments/files/17021319/glassfish.man.txt)

Demo:

https://github.com/user-attachments/assets/404d6b30-9271-447a-840a-bea61a89ed7c

OndroMih committed:
Runnable GF Embedded: Initial prototype

Example of running an app from command line:

java -jar glassfish-embedded-all.jar "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.enabled=true" "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.port=8080" app.war

Example with root context:

java -jar glassfish-embedded-all.jar "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.enabled=true" "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.port=8080" "deploy --contextroot=/ app.war"

OndroMih committed:
Runnable GF Embedded: Initial prototype

Example of running an app from command line:

java -jar glassfish-embedded-all.jar "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.enabled=true" "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.port=8080" app.war

Example with root context:

java -jar glassfish-embedded-all.jar "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.enabled=true" "set configs.config.server-config.network-config.network-listeners.network-listener.http-listener.port=8080" "deploy --contextroot=/ app.war"

OndroMih committed:
Runnable GF Embedded: Support setting properties and app param

Deploy app if the argument is a file or dir.
If it's a single app, deploy to root context.
Support --option=value arguments with values that set GlassFish properties
Support several special options, like --port, --properties.
Support loading properties from glassfish.properties in the current directory.

Revert using FallbackStartupModule, was needed only for testing

OndroMih committed:
Runnable GF Embedded: Support setting properties and app param

Deploy app if the argument is a file or dir.
If it's a single app, deploy to root context.
Support --option=value arguments with values that set GlassFish properties
Support several special options, like --port, --properties.
Support loading properties from glassfish.properties in the current directory.

Revert using FallbackStartupModule, was needed only for testing

OndroMih committed:
Runnable GF Embedded: Print help

OndroMih committed:
Runnable GF Embedded: Print help

OndroMih committed:
Runnable GF Embedded: Commands in properties file

OndroMih committed:
Runnable GF Embedded: Commands in properties file

OndroMih committed:
Runnable GF Embedded: Applications in properties file

OndroMih committed:
Runnable GF Embedded: Applications in properties file

OndroMih committed:
Runnable GF Embedded: Autodeploy directory option

Refactoring the API, so that UberMain can be configured and executed programmatically.

OndroMih committed:
Runnable GF Embedded: Autodeploy directory option

Refactoring the API, so that UberMain can be configured and executed programmatically.

OndroMih committed:
Runnable GF Embedded: Fixes in word wrapper for help output

OndroMih committed:
Runnable GF Embedded: Fixes in word wrapper for help output

OndroMih committed:
Runnable GF Embedded: Default domain directory and config file

OndroMih committed:
Runnable GF Embedded: Default domain directory and config file

OndroMih committed:
Runnable GF Embedded: Documentation

OndroMih committed:
Runnable GF Embedded: Documentation

OndroMih committed:
Runnable GF Embedded: Logging configuration

OndroMih committed:
Runnable GF Embedded: Logging configuration

OndroMih committed:
Runnable GF Embedded: Print info after startup

Like:

Refactored to the kernel module which already contains the config-api dependency.
Adding config-api to the bootstrap module caused a dependency cycle that failed the Maven build.

OndroMih committed:
Runnable GF Embedded: Print info after startup

Like:

Refactored to the kernel module which already contains the config-api dependency.
Adding config-api to the bootstrap module caused a dependency cycle that failed the Maven build.

OndroMih committed:
Runnable GF Embedded: Make the Static shell JAR runnable

Add some missing dependencies that are outside of the default directory.

This makes Static shell equal to GF Embedded - it can be added to
the classpath or used as an executable JAR instead of GF Embedded JAR.

Fix WebSockets in GF Static Shell

OndroMih committed:
Runnable GF Embedded: Make the Static shell JAR runnable

Add some missing dependencies that are outside of the default directory.

This makes Static shell equal to GF Embedded - it can be added to
the classpath or used as an executable JAR instead of GF Embedded JAR.

Fix WebSockets in GF Static Shell

OndroMih committed:
Runnable GF Embedded: Fixes for Embedded Web edition

Added missing Jakarta APIs:
- Validation is part of Web Profile
- XML Binding is not part of Web Profile but JobManagerService depends on it

Improvements in handling exceptions and logging messages.

OndroMih committed:
Runnable GF Embedded: Fixes for Embedded Web edition

Added missing Jakarta APIs:
- Validation is part of Web Profile
- XML Binding is not part of Web Profile but JobManagerService depends on it

Improvements in handling exceptions and logging messages.

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/UnknownPropertyException.java:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author Ondro Mihalyi
+ */
+public class UnknownPropertyException extends Exception {
```
Can this be final?

OndroMih replied:
Done

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/UnknownPropertyException.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public class UnknownPropertyException extends Exception {
+
+    private String key;
```
Can these be final?
This would help to show they are never assigned.

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/UnknownPropertyException.java:
```diff
@@ -0,0 +1,44 @@
+        return key;
+    }
+
+    public String getValue() {
```
unused

OndroMih replied:
Removed.

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/WordWrapper.java:
```diff
@@ -0,0 +1,108 @@
+ */
+public class WordWrapper {
+
+    final int MAX_LINE_LENGTH;
```
I'd rather have them private.

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/WordWrapper.java:
```diff
@@ -0,0 +1,108 @@
+
+    final int MAX_LINE_LENGTH;
+    final String HELP_LINE_INDENT;
+    int characterCount;
```
I'd rather have it as local.

OndroMih replied:
This is an instance field, not a local variable. It's initialized based on constructor argument `initialIndent`.

OndroMih replied:
This must be an instance field, not a local variable. It holds the state. I refactored this class into a direct collector. That should make it clear it's a stateful class and shouldn't be reused.

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/WordWrapper.java:
```diff
@@ -0,0 +1,108 @@
+ *
+ * @author Ondro Mihalyi
+ */
+public class WordWrapper {
```
I'd like to to be final.

OndroMih replied:
Done

pzygielo commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/WordWrapper.java:
```diff
@@ -0,0 +1,108 @@
+        return sb.toString();
+    }
+
+    public Collector<String, StringBuilder, String> collector() {
```
Seems to be used only in test. I'd like it to be removed from here.

OndroMih replied:
It shouldn't be used only in tests. I forgot to apply it in the production code. Fixed now.

OndroMih committed:
Runnable GF Embedded: Add option to shut down after initialized

OndroMih committed:
Runnable GF Embedded: Add option to shut down after initialized

OndroMih committed:
Runnable GF Embedded: Update copyright info

OndroMih committed:
Runnable GF Embedded: Update copyright info

OndroMih committed:
Runnable GF Embedded: Cleanup based on review comments

Turned WordWraper into a direct collector WordWrapCollector.
Moved the man page into the proper module and location.
INFO level for output from admin commands

OndroMih committed:
Runnable GF Embedded: Cleanup based on review comments

Turned WordWraper into a direct collector WordWrapCollector.
Moved the man page into the proper module and location.
INFO level for output from admin commands

dmatej commented on docs/embedded-server-guide/src/main/asciidoc/embedded-server-guide.adoc:
```diff
@@ -173,39 +382,10 @@ Embedded Server API]
+=== Setting the Class Path
 
 To enable your applications to locate the class libraries for embedded
-{productName}, add one of the following JAR files to your class path:
```
I think you can link to Maven Central here.

dmatej replied:
Aha, you already did that and moved the list up, ok, perfect!

dmatej commented on nucleus/hk2-config-generator/src/main/java/org/jvnet/hk2/config/ConfigParser.java:
```diff
@@ -140,7 +142,9 @@ protected Dom handleElement(XMLStreamReader in,DomDocument document, Dom parent)
-            Logger.getAnonymousLogger().severe("Ignoring unrecognized element "+in.getLocalName() + " at " + in.getLocation());
+            logger.log(Level.WARNING, () -> "Ignoring unrecognized element " + in.getLocalName() + " at line #" + in.getLocation().getLineNumber());
+            logger.log(Level.FINE, () -> "Unrecognized element " + in.getLocalName() + " at "
+                    + in.getLocation().toString().replace("\n", ", "));
```
I am not sure if it is good to manipulate the location toString here, but ... or I will say it. It is wrong, but I have found that `EventLocation` (`jaxb-osgi`) really prints `\n` character after every value ... perhaps we could create a PR for them ;-)

dmatej commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/Option.java:
```diff
@@ -0,0 +1,264 @@
+            setLogLevel(logLevel, arguments);
+        }
+    },
+    LOG_PROPERTIES("logProperties", "--logProperties=FILE",
```
This will cause issues, because JDK loads first and also starts using the logging configuration, the only reliable option is to use -Djava.util.logging.config.file on command line.

dmatej commented on nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/commandline/Option.java:
```diff
@@ -0,0 +1,264 @@
+        for (Handler handler : rootLogger.getHandlers()) {
+            handler.setLevel(level);
+        }
+        LogManager.getLogManager().getLoggerNames().asIterator()
```
Problem with this is that it returns just already initialized loggers. It would be better to retrieve current configuration from the logmanager, change it, and then call logManager.readConfiguration. Or just provide our own defaults generated using this parameter, set the root logger+handler level and just call logger.readConfiguration asap, even before main (in static init). It would be done just if user would not explicitly set the `-Djava.util.logging.config.file` or `-Djava.util.logging.config.class.`

In the video it seems that currently it uses JDK defaults.

dmatej commented:
I have rebased the branch and resolved conflicts as it was conflicting with my changes in master - finally they were just minor.

OndroMih committed:
Merge my/ondromih-runnable-glassfish-embedded into ondromih-runnable-glassfish-embedded

# Conflicts:
#	nucleus/core/kernel/src/main/java/org/glassfish/runnablejar/UberMain.java

OndroMih committed:
Runnable GF Embedded: Add Add-Opens and Add-Exports to MANIFEST

These match the configuration in the default GlassFish server domain.xml. 
With this it's no longer needed to specify --add-opens and --add-exports on the command line.

OndroMih committed:
Runnable GF Embedded: Document running with java -cp
